cpu/qn908x/periph/spi.c

```diff
@@ -0,0 +1,341 @@
+    (((1u << SPI_HWCS_NUMOF) - 1) << SPI_FIFOWR_TXSSEL0_N_SHIFT)
+
+/**
+ * @brief Initialize a SPI transfer given the transfer parameters.
+ */
+static void _spi_config_transfer(spi_pending_transfer_t *tr, spi_cs_t cs,
+                                 bool cont, const void *out, void *in,
+                                 size_t len)
+{
+    tr->in = in;
+    tr->in_len = in ? len : 0;
+    tr->out = out;
+    tr->tr_len = len;
+    tr->tx_mask = SPI_HWCS_DEASSERT_ALL;
+    if (GPIO_T_IS_HWCS(cs)) {
+        /* Flag that the TX should assert this HWCS by clearing the bit. */
+        tr->tx_mask &= ~(1u << (SPI_FIFOWR_TXSSEL0_N_SHIFT + GPIO_T_HWCS(cs)));
+        if (!cont) {
+            /* Flag the End of Transfer (EOT) in the mask. This will only be
+             * used in the last byte. */
+            tr->tx_mask |= SPI_FIFOWR_EOT_MASK;
+        }
+    }
+    if (!in) {
+        /* Ignores the RX side when the @p in is NULL so we don't need to read
+         * the FIFO at all. */
+        tr->tx_mask |= SPI_FIFOWR_RXIGNORE_MASK;
+    }
+    tr->tx_mask |= SPI_FIFOWR_LEN(7);  /* Data transfers of 8 bits. */
+}
+
+/**
+ * @brief Perform a blocking SPI transfer.
+ */
+static void _spi_transfer_blocking(spi_t bus, spi_pending_transfer_t *tr)
+{
+    SPI_Type *const spi_bus = spi_config[bus].dev;
+
+    /* Configure to use the RX and TX fifo, and empty them. */
+    spi_bus->FIFOCFG = SPI_FIFOCFG_ENABLETX_MASK |
+                       SPI_FIFOCFG_ENABLERX_MASK |
+                       SPI_FIFOCFG_EMPTYTX_MASK | SPI_FIFOCFG_EMPTYRX_MASK;
+    spi_bus->FIFOSTAT = SPI_FIFOSTAT_TXERR_MASK | SPI_FIFOSTAT_RXERR_MASK;
+
+    while (tr->in_len || tr->tr_len) {
+        /* Read from RX FIFO if possible. */
+        if (spi_bus->FIFOSTAT & SPI_FIFOSTAT_RXNOTEMPTY_MASK) {
+            uint32_t rd = spi_bus->FIFORD;
+            if (tr->in_len) {
+                *(tr->in++) = (uint8_t)rd;
+                tr->in_len--;
+            }
+        }
+
+        /* Write when able to write and we have data to send or bogus (0) bytes
+         * to send when in receive-only mode. */
+        if ((spi_bus->FIFOSTAT & SPI_FIFOSTAT_TXNOTFULL_MASK) && tr->tr_len) {
+            uint32_t wr = tr->tx_mask;
+            if (tr->out) {
+                wr |= *(tr->out++);
+            }
+
+            /* If this is *not* the last byte, remove the EOT flag if any. */
+            tr->tr_len--;
+            if (tr->tr_len) {
+                wr &= ~SPI_FIFOWR_EOT_MASK;
+            }
+            /* Push the data to the FIFO. */
+            spi_bus->FIFOWR = wr;
+        }
+    }
+    _spi_wait_txempty(spi_bus);
+}
+
+void spi_transfer_bytes(spi_t bus, spi_cs_t cs, bool cont,
+                        const void *out, void *in, size_t len)
+{
+    spi_pending_transfer_t tr;
+
+    _spi_config_transfer(&tr, cs, cont, out, in, len);
+
+    /* At least one of input or one output buffer is given */
+    assert(bus < SPI_NUMOF);
+
+    if (!GPIO_T_IS_HWCS(cs)) {
```

Comment:
You could also do this in `_spi_config_transfer()` so it's all in one place.

Reply:
The gpio_set() that matches this gpio_clear() is in this function, that's why I placed it here instead of `_spi_config_transfer`.